File: problems/56.合并区间.py
#
# @lc app=leetcode.cn id=56 lang=python3
#
# [56] 合并区间
#

# @lc code=start
from cProfile import label


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key = lambda x:x[0])
        result = []
        for i in range(len(intervals)):
            if not result or result[-1][-1] < intervals[i][0]:#只有结果数组为空或者没有交集的情况才直接加入
                result.append(intervals[i])
            else:
                result[-1] = [result[-1][0],max(intervals[i][-1],result[-1][-1])]#取两个数组中较大的数作为上界
        return result

# @lc code=end

File: problems/56.合并区间.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;


/*
 * @lc app=leetcode.cn id=56 lang=java
 *
 * [56] 合并区间
 */

// @lc code=start
class Solution {
    public int[][] merge(int[][] intervals) {
        if(intervals.length == 0)
        return new int[0][2];

        Arrays.sort(intervals, new Comparator<int[]>(){
            public int compare(int[] a, int b[]){
                return a[0] - b[0];
            }
        });
    
        // 
        List<int[]> res = new ArrayList<>();
        for(int i = 0 ; i < intervals.length ;){
            int right = intervals[i][1];
            int j = i + 1;
            while(j<intervals.length && intervals[j][0] <= right){
                right = Math.max(right, intervals[j][1]);
                j++;
            }
            
            res.add(new int[] {intervals[i][0],right});
            i = j;
        }
    
    // 处理最后一个

    return res.toArray(new int[res.size()][]);
    }
    
}
// @lc code=end

